src/katgpucbf/xbgpu/engine.py

```diff
@@ -495,6 +854,39 @@ def populate_sensors(self, sensors: aiokatcp.SensorSet, rx_sensor_timeout: float
         self.add_service_task(time_sync_task)
         self._cancel_tasks.append(time_sync_task)
 
+    def free_rx_item(self, item: RxQueueItem) -> None:
+        """Return an RxQueueItem to the free queue if its refcount hits zero."""
+        item.refcount -= 1
+        if item.refcount == 0:
+            # All Pipelines are done with this item
+            self._active_in_sem.release()
+            # Need to wait for any outstanding events before recycling
+            self.add_service_task(
+                asyncio.create_task(
+                    self._push_recv_chunk(item),
+                    name="Receive Chunk Recycle Task",
+                )
+            )
+            item.reset()
+            self._rx_free_item_queue.put_nowait(item)
+
+    @staticmethod
+    async def _push_recv_chunk(item: RxQueueItem) -> None:
+        """Return chunk to the stream once `events` have fired."""
+        await item.async_wait_for_events()
+        item.chunk.recycle()  # type: ignore
+        item.chunk = None
+
+    async def _add_rx_item(self, item: RxQueueItem) -> None:
+        """Push an :class:`RxQueueItem` to all the pipelines.
+
+        Allow for None to be passed through to indicate shutdown procedure.
```

Comment:
Is this meant to be a TODO? It doesn't match the type annotation or the body of the function.

Reply:
Ah, legacy from my own reminder (now done with the `shutdown`). Will remove.